### ISSUE TITLE
Windows build (Visual Studio 2019)

### ISSUE THREAD

**Issue by NPTP** [dilevin/computer-graphics-mass-spring-systems#13]
Hi,

For the past few assignments (since libigl was introduced) I haven't been able to get any builds working at home on Windows (Win7 x64, Visual Studio 2019), and have had to work on them exclusively in the CDF labs. Of course the labs are now closed so I need to find a way, but I haven't been able to manage it.

On a fresh recursive clone of the repo, I invariably get the following errors when trying to build (same errors for dense/sparse executables):
![image](https://user-images.githubusercontent.com/16977354/76995897-efbc7300-6926-11ea-9047-315764796fa3.png)

I've been searching online for a solution but nothing has worked yet. If anyone has any ideas or pointers to things that helped them, let me know. Thanks!

**Comment by abhimadan** [comment]
Strange, it looks like it's not linking in references from the standard library. Can you try making a simple VS project that uses basic string functions and see if that compiles?

**Comment by abhimadan** [comment]
Also, another TA pointed out that this could be due to a conflicting GLFW installation in your path. Make sure that libigl is linking with the GLFW library built locally.

**Comment by rarora7777** [comment]
If you don't know how, here are the steps:
In Visual Studio, go the the `glfw` project and look at the `Source Files` for `glfw`. Let us know where those files are located. If it's inside the assignment repo, then it's ok. Otherwise, you need to remove your existing GLFW installation from the system path.

**Comment by rarora7777** [comment]
We regularly build and test all assignments on Windows 10, MacOS, and Ubuntu. It is expected to work on all common systems.

**Comment by NPTP** [comment]
Thanks for everyone pitching in to help with this!

> Strange, it looks like it's not linking in references from the standard library. Can you try making a simple VS project that uses basic string functions and see if that compiles?

I can build such a project (using <cstring>) with 0 errors. Also, I was able to build and run all the assignments with the same setup, up until A5. Rather than spend time trying to fix it, I just started working in the labs instead.

> Make sure that libigl is linking with the GLFW library built locally.

How might I go about this?

> In Visual Studio, go the the `glfw` project 

Hi, thanks for the tips. Do you just mean here?
![image](https://user-images.githubusercontent.com/16977354/77003646-1f25ac80-6934-11ea-94dc-2318e563278e.png)
If so, these are the only options I get on a right-click:
![image](https://user-images.githubusercontent.com/16977354/77003899-8c394200-6934-11ea-90ab-19e98b27daba.png)
Not sure how to find the `glfw` "project" and `Source Files`!

Thanks!

**Comment by rarora7777** [comment]
That confirms my assumption that cmake is finding a pre-installed glfw on your system and using that. 
Let's try this fix:
1. Open up `./CMakeLists.txt` and add this line after `cmake_minimum_required(VERSION 3.1)`:
`set(MSVC_RUNTIME "dynamic")`
2. Delete `CMakeCache.txt` from `./build/`.
3. Run `cmake ..` and then rebuild from VS.

**Comment by rarora7777** [comment]
If this doesn't work, then we can try other fixes to try to make sure that cmake doesn't use the pre-installed GLFW.

**Comment by NPTP** [comment]
> That confirms my assumption that cmake is finding a pre-installed glfw on your system and using that.
> Let's try this fix:

That did it! Thanks so much for your help! I've really gotta get a handle on Visual Studio...

Interestingly, the right-click menu I showed you hasn't changed at all since making the fix you suggested, even though the build is now working.

**Comment by rarora7777** [comment]
Yeah, that's because the fix I gave still doesn't use the GLFW included with libigl. It's just making sure the rest of the assignment is linked dynamically with the C Runtime library. Your original screenshot suggested that the GLFW you were using was being linked dynamically with the C Runtime. Normally, libigl forces anything using it to link dynamically with the C Runtime on Windows. This is needed for a component of libigl that we never use for this course. Since we never use that component, it's perfectly acceptable to link the runtime dynamically.

**Comment by NPTP** [comment]
I see. Thanks again!

**Comment by dilevin** [comment]
just leaving this open in case someone else has this problem
